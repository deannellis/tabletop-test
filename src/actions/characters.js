import uuid from 'uuid';

// ADD_CHARACTER
export const addCharacter = (
    {
        abilities = {
            strength: 0,
            intelligence: 0,
            wisdom: 0,
            dexterity: 0,
            constitution: 0,
            charisma: 0
        }, 
        race = '', 
        charClass = '', 
        hp = 0, 
        xp = 0,
        name = '',
        description = ''
    } = {}
) => ({
    type: 'ADD_CHARACTER',
    character: {
        id: uuid(),
        abilities,
        race,
        charClass,
        hp,
        xp,
        name,
        description
    }
});

// REMOVE_CHARACTER

export const removeCharacter = ({ id } = {}) => ({
    type: 'REMOVE_CHARACTER',
    id
});

// EDIT_CHARACTER
export const editCharacter = (id, updates) => ({
    type: 'EDIT_CHARACTER',
    id,
    updates
});

