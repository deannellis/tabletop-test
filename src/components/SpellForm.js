import React from 'react';
import spells from '../objects/spells';
import SpellListItem from './SpellListItem';
import { Link } from 'react-router-dom';

class SpellForm extends React.Component {
    constructor(props){
        super(props);
        
        this.state = {
            selectedOption: '',
            error: ''
        };
    }
    
    render () {
        if (this.props.currentCharacter !== undefined) {
            return (
                <div>
                    <h2>Choose a Spell</h2>
                    <form onSubmit={this.onSubmit}>
                        {spells.map((spell) => {
                            return <SpellListItem 
                                        key={spell.id} 
                                        {...spell}
                                        spellType={'magic-user'}
                                        selected={this.state.selectedOption}
                                        handleSelectRadio={this.handleSelectRadio}
                                    />
                        })}
                        <button>Next Step</button>
                    </form>
                </div>    
            )
        } else{
            return (
                <div>
                    Character not found
                    <div>
                        <Link to="/">Home</Link>
                    </div>
                </div>    
            )
        }
    }
    
    handleSelectRadio = (selection) => {
        this.setState(() => ({
            selectedOption: selection
        }));
    }
    
    onSubmit = (e) => {
        e.preventDefault();
        if(!this.state.selectedOption) {
            this.setState(() => ({
                error: 'Please select a spell'
            }));
        } else {
            this.setState(() => ({
                error: ''
            }));
            this.props.onSubmit( this.props.currentCharacter.id, {
                spells: [this.state.selectedOption],
                inProgressStep: 5
            });
        }
    };
}

export default SpellForm;