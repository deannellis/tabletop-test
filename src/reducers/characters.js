//  CHARACTERS REDUCER
import actionIds from 'actions/actionIds';
const {ADD_CHARACTER, REMOVE_CHARACTER, EDIT_CHARACTER, COMPLETE_CHARACTER} = actionIds;
const charactersReducerDefaultState = [];

export default ( state = charactersReducerDefaultState, action ) => {
    switch (action.type) {
        case ADD_CHARACTER:
            return [
                ...state,
                action.character
            ];
        case REMOVE_CHARACTER:
            return state.filter(({ id }) => {
                return id !== action.id;
            });
        case EDIT_CHARACTER:
            return state.map((character) => {
                if (character.id ===  action.id) {
                    return {
                        ...character,
                        ...action.updates
                    }
                } else {
                    return character;
                }
            });
        case COMPLETE_CHARACTER:
            return state.map((character) => {
                if (character.id ===  action.id) {
                    let {inProgressStep, ...rest} = character;
                    console.log('rest'+rest);
                    return rest;
                } else {
                    return character;
                }
            });
        default:
            return state;
    }
};