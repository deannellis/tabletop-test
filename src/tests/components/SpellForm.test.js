/* global jest expect */
import React from 'react';
import { shallow } from 'enzyme';
import SpellForm from '../../components/SpellForm';
import characters from '../fixtures/characters';

test('should render class form without props (character not found)', () => {
    const wrapper = shallow(<SpellForm />);
    expect(wrapper).toMatchSnapshot();
});

test('should render spell form with props', () => {
    const wrapper = shallow(<SpellForm currentCharacter={characters[0]}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid submission', () => {
    const wrapper = shallow(<SpellForm currentCharacter={characters[1]}/>);
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    });
    expect(wrapper.state('error').length).toBeGreaterThan(0);
    expect(wrapper).toMatchSnapshot();
});

test('should call onSubmit prop for valid submission', () => {
    const onSubmitSpy = jest.fn();
    const wrapper = shallow(<SpellForm currentCharacter={characters[1]} onSubmit={onSubmitSpy} />);
    const value = 0;
    wrapper.setState({ selectedOption: value });
    wrapper.find('form').simulate('submit', {
        preventDefault: () => {}
    });
    expect(wrapper.state('error')).toBe('');
    expect(onSubmitSpy).toHaveBeenLastCalledWith(
        characters[1].id,
        {
            spells: [value],
            inProgressStep: 5
        }
    );
});