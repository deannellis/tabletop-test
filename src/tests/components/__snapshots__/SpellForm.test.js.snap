// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render class form without props (character not found) 1`] = `<NotFoundPage />`;

exports[`should render error for invalid submission 1`] = `
<div>
  <h2>
    Choose a Spell
  </h2>
  <p>
    Please select a spell
  </p>
  <form
    onSubmit={[Function]}
  >
    <Component
      duration="special"
      handleSelectRadio={[Function]}
      id={0}
      key="0"
      level={1}
      longDescription="This spell makes a humanoid creature of 4 hit dice or less regard the caster as its trusted friend and ally. And more..."
      name="Charm Person"
      range="30ft"
      selected={null}
      shortDescription="This spell makes a humanoid creature regard the caster as its trusted friend and ally."
      spellType="magic-user"
      type="magic-user"
    />
    <Component
      duration="2 turns"
      handleSelectRadio={[Function]}
      id={1}
      key="1"
      level={1}
      longDescription="The caster of this spell is able to detect enchanted or enspelled objects or creatures within the given range by sight, seeing them surrounded by a pale glowing light. And more..."
      name="Detect Magic"
      range="60ft"
      selected={null}
      shortDescription="The caster of this spell is able to detect magic objects or creatures."
      spellType="magic-user"
    />
    <Component
      duration="5 turns + 1/level"
      handleSelectRadio={[Function]}
      id={2}
      key="2"
      level={1}
      longDescription="This spell creates an invisible, slightly concave circular plane of force for carrying loads. And more..."
      name="Floating Disc"
      range="10ft"
      selected={null}
      shortDescription="This spell creates an invisible plane of force for carrying loads."
      spellType="magic-user"
      type="magic-user"
    />
    <Component
      duration="1 round/level"
      handleSelectRadio={[Function]}
      id={3}
      key="3"
      level={1}
      longDescription="This spell magically holds shut a door, gate, window, or shutter of wood, metal, or stone. And more..."
      name="Hold Portal"
      range="100ft + 10ft/level"
      selected={null}
      shortDescription="This spell magically holds shut a portal (door, gate, window, etc)"
      spellType="magic-user"
      type="magic-user"
    />
    <Component
      duration="6 turns + 1/level"
      handleSelectRadio={[Function]}
      id={4}
      key="4"
      level={1}
      longDescription="This spell creates a light equal to torchlight which illuminates a 30ft radius area (and provides dim light for an additional 20) around the target location or object. And more..."
      name="Light"
      range="120ft"
      selected={null}
      shortDescription="This spell creates a light equal to torchlight"
      spellType="magic-user"
    />
    <Component
      duration="instantaneous"
      handleSelectRadio={[Function]}
      id={5}
      key="5"
      level={1}
      longDescription="With this spell the caster heals 1d6+1 hit points of damage by laying his or her hand upon the injured creature."
      name="Cure Light Wounds"
      range="touch"
      selected={null}
      shortDescription="The caster heals 1d6+1 hit points of damage"
      spellType="magic-user"
      type="cleric"
    />
    <button>
      Next Step
    </button>
  </form>
</div>
`;

exports[`should render spell form with props 1`] = `
<div>
  <h2>
    Choose a Spell
  </h2>
  <form
    onSubmit={[Function]}
  >
    <Component
      duration="special"
      handleSelectRadio={[Function]}
      id={0}
      key="0"
      level={1}
      longDescription="This spell makes a humanoid creature of 4 hit dice or less regard the caster as its trusted friend and ally. And more..."
      name="Charm Person"
      range="30ft"
      selected={null}
      shortDescription="This spell makes a humanoid creature regard the caster as its trusted friend and ally."
      spellType="magic-user"
      type="magic-user"
    />
    <Component
      duration="2 turns"
      handleSelectRadio={[Function]}
      id={1}
      key="1"
      level={1}
      longDescription="The caster of this spell is able to detect enchanted or enspelled objects or creatures within the given range by sight, seeing them surrounded by a pale glowing light. And more..."
      name="Detect Magic"
      range="60ft"
      selected={null}
      shortDescription="The caster of this spell is able to detect magic objects or creatures."
      spellType="magic-user"
    />
    <Component
      duration="5 turns + 1/level"
      handleSelectRadio={[Function]}
      id={2}
      key="2"
      level={1}
      longDescription="This spell creates an invisible, slightly concave circular plane of force for carrying loads. And more..."
      name="Floating Disc"
      range="10ft"
      selected={null}
      shortDescription="This spell creates an invisible plane of force for carrying loads."
      spellType="magic-user"
      type="magic-user"
    />
    <Component
      duration="1 round/level"
      handleSelectRadio={[Function]}
      id={3}
      key="3"
      level={1}
      longDescription="This spell magically holds shut a door, gate, window, or shutter of wood, metal, or stone. And more..."
      name="Hold Portal"
      range="100ft + 10ft/level"
      selected={null}
      shortDescription="This spell magically holds shut a portal (door, gate, window, etc)"
      spellType="magic-user"
      type="magic-user"
    />
    <Component
      duration="6 turns + 1/level"
      handleSelectRadio={[Function]}
      id={4}
      key="4"
      level={1}
      longDescription="This spell creates a light equal to torchlight which illuminates a 30ft radius area (and provides dim light for an additional 20) around the target location or object. And more..."
      name="Light"
      range="120ft"
      selected={null}
      shortDescription="This spell creates a light equal to torchlight"
      spellType="magic-user"
    />
    <Component
      duration="instantaneous"
      handleSelectRadio={[Function]}
      id={5}
      key="5"
      level={1}
      longDescription="With this spell the caster heals 1d6+1 hit points of damage by laying his or her hand upon the injured creature."
      name="Cure Light Wounds"
      range="touch"
      selected={null}
      shortDescription="The caster heals 1d6+1 hit points of damage"
      spellType="magic-user"
      type="cleric"
    />
    <button>
      Next Step
    </button>
  </form>
</div>
`;
